File: frontend/src/core/dynamicConfig.ts
/**
 * Dinamik Konfigürasyon Sistemi
 * Runtime'da değiştirilebilir ayarlar
 */

import { apiClient } from './apiClient';
import { logger } from './logger';
import { eventBus } from './eventBus';

export interface DynamicConfigItem {
  key: string;
  value: any;
  category: string;
  description: string;
  dataType: string;
  isSensitive?: boolean;
  isReadonly?: boolean;
  minValue?: any;
  maxValue?: any;
  allowedValues?: any[];
  updatedAt: string;
  updatedBy?: string;
}

export interface DynamicConfigCategory {
  [key: string]: {
    value: any;
    description: string;
    dataType: string;
    isSensitive?: boolean;
    isReadonly?: boolean;
    updatedAt: string;
    updatedBy?: string;
  };
}

export type ConfigChangeListener = (key: string, oldValue: any, newValue: any) => void;

class DynamicConfigManager {
  private config: Map<string, DynamicConfigItem> = new Map();
  private listeners: ConfigChangeListener[] = [];
  private cache: Map<string, { value: any; timestamp: number }> = new Map();
  private cacheTimeout = 300000; // 5 minutes

  constructor() {
    this.setupEventListeners();
    this.loadConfig();
  }

  private setupEventListeners(): void {
    // Listen for config changes from other tabs
    window.addEventListener('storage', (event) => {
      if (event.key === 'dynamic_config_update') {
        const { key, oldValue, newValue } = JSON.parse(event.newValue || '{}');
        this.handleConfigChange(key, oldValue, newValue);
      }
    });

    // Listen for online/offline events
    window.addEventListener('online', () => {
      this.loadConfig();
    });
  }

  private async loadConfig(): Promise<void> {
    try {
      const response = await apiClient.get('/api/config/dynamic');
      const configData = response.data;

      this.config.clear();
      Object.entries(configData).forEach(([key, item]: [string, any]) => {
        this.config.set(key, item as DynamicConfigItem);
      });

      logger.info('Dynamic config loaded', { count: this.config.size });
      eventBus.emit('config:loaded', { count: this.config.size });
    } catch (error) {
      logger.error('Failed to load dynamic config', { error });
      this.loadFromCache();
    }
  }

  private loadFromCache(): void {
    try {
      const cached = localStorage.getItem('dynamic_config');
      if (cached) {
        const configData = JSON.parse(cached);
        this.config.clear();
        Object.entries(configData).forEach(([key, item]: [string, any]) => {
          this.config.set(key, item as DynamicConfigItem);
        });
        logger.info('Dynamic config loaded from cache', { count: this.config.size });
      }
    } catch (error) {
      logger.error('Failed to load config from cache', { error });
    }
  }

  private saveToCache(): void {
    try {
      const configData: { [key: string]: DynamicConfigItem } = {};
      this.config.forEach((item, key) => {
        configData[key] = item;
      });
      localStorage.setItem('dynamic_config', JSON.stringify(configData));
    } catch (error) {
      logger.error('Failed to save config to cache', { error });
    }
  }

  public get(key: string, defaultValue: any = null): any {
    // Check cache first
    const cached = this.cache.get(key);
    if (cached && Date.now() - cached.timestamp < this.cacheTimeout) {
      return cached.value;
    }

    // Get from config
    const item = this.config.get(key);
    if (item) {
      // Update cache
      this.cache.set(key, { value: item.value, timestamp: Date.now() });
      return item.value;
    }

    return defaultValue;
  }

  public async set(key: string, value: any, updatedBy?: string): Promise<boolean> {
    try {
      const response = await apiClient.put('/api/config/dynamic', {
        key,
        value,
        updatedBy,
      });

      if (response.status === 200) {
        const oldValue = this.get(key);
        const newItem: DynamicConfigItem = {
          key,
          value,
          category: response.data.category || 'general',
          description: response.data.description || '',
          dataType: response.data.dataType || 'string',
          isSensitive: response.data.isSensitive || false,
          isReadonly: response.data.isReadonly || false,
          updatedAt: new Date().toISOString(),
          updatedBy,
        };

        this.config.set(key, newItem);
        this.cache.set(key, { value, timestamp: Date.now() });
        this.saveToCache();

        // Notify listeners
        this.notifyListeners(key, oldValue, value);

        // Broadcast to other tabs
        localStorage.setItem('dynamic_config_update', JSON.stringify({
          key,
          oldValue,
          newValue: value,
        }));
        localStorage.removeItem('dynamic_config_update');

        logger.info('Dynamic config updated', { key, value });
        eventBus.emit('config:updated', { key, oldValue, newValue: value });

        return true;
      }

      return false;
    } catch (error) {
      logger.error('Failed to update dynamic config', { key, value, error });
      return false;
    }
  }

  public getCategory(category: string): DynamicConfigCategory {
    const result: DynamicConfigCategory = {};
    
    this.config.forEach((item, key) => {
      if (item.category === category) {
        result[key] = {
          value: item.value,
          description: item.description,
          dataType: item.dataType,
          isSensitive: item.isSensitive,
          isReadonly: item.isReadonly,
          updatedAt: item.updatedAt,
          updatedBy: item.updatedBy,
        };
      }
    });

    return result;
  }

  public getAll(): { [key: string]: DynamicConfigItem } {
    const result: { [key: string]: DynamicConfigItem } = {};
    this.config.forEach((item, key) => {
      result[key] = item;
    });
    return result;
  }

  public getKeys(): string[] {
    return Array.from(this.config.keys());
  }

  public getKeysByCategory(category: string): string[] {
    const keys: string[] = [];
    this.config.forEach((item, key) => {
      if (item.category === category) {
        keys.push(key);
      }
    });
    return keys;
  }

  public has(key: string): boolean {
    return this.config.has(key);
  }

  public addListener(listener: ConfigChangeListener): () => void {
    this.listeners.push(listener);
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index > -1) {
        this.listeners.splice(index, 1);
      }
    };
  }

  private notifyListeners(key: string, oldValue: any, newValue: any): void {
    this.listeners.forEach(listener => {
      try {
        listener(key, oldValue, newValue);
      } catch (error) {
        logger.error('Config listener error', { key, error });
      }
    });
  }

  private handleConfigChange(key: string, oldValue: any, newValue: any): void {
    // Update local config
    const item = this.config.get(key);
    if (item) {
      item.value = newValue;
      item.updatedAt = new Date().toISOString();
      this.config.set(key, item);
      this.cache.set(key, { value: newValue, timestamp: Date.now() });
    }

    // Notify listeners
    this.notifyListeners(key, oldValue, newValue);
  }

  public async refresh(): Promise<void> {
    await this.loadConfig();
  }

  public clearCache(): void {
    this.cache.clear();
  }

  public getStats(): {
    totalConfigs: number;
    categories: string[];
    cacheSize: number;
    listeners: number;
  } {
    const categories = new Set<string>();
    this.config.forEach(item => {
      categories.add(item.category);
    });

    return {
      totalConfigs: this.config.size,
      categories: Array.from(categories),
      cacheSize: this.cache.size,
      listeners: this.listeners.length,
    };
  }

  public exportConfig(): string {
    return JSON.stringify(this.getAll(), null, 2);
  }

  public importConfig(configData: string): boolean {
    try {
      const data = JSON.parse(configData);
      Object.entries(data).forEach(([key, item]: [string, any]) => {
        this.config.set(key, item as DynamicConfigItem);
      });
      this.saveToCache();
      return true;
    } catch (error) {
      logger.error('Failed to import config', { error });
      return false;
    }
  }

  // Convenience methods for common config types
  public getString(key: string, defaultValue: string = ''): string {
    const value = this.get(key, defaultValue);
    return typeof value === 'string' ? value : String(value);
  }

  public getNumber(key: string, defaultValue: number = 0): number {
    const value = this.get(key, defaultValue);
    return typeof value === 'number' ? value : Number(value);
  }

  public getBoolean(key: string, defaultValue: boolean = false): boolean {
    const value = this.get(key, defaultValue);
    return typeof value === 'boolean' ? value : Boolean(value);
  }

  public getArray(key: string, defaultValue: any[] = []): any[] {
    const value = this.get(key, defaultValue);
    return Array.isArray(value) ? value : [];
  }

  public getObject(key: string, defaultValue: object = {}): object {
    const value = this.get(key, defaultValue);
    return typeof value === 'object' && value !== null ? value : defaultValue;
  }
}

// Singleton instance
export const dynamicConfig = new DynamicConfigManager();

// Export commonly used methods
export const {
  get: getDynamicConfig,
  set: setDynamicConfig,
  getCategory: getDynamicConfigCategory,
  getAll: getAllDynamicConfig,
  getString: getDynamicConfigString,
  getNumber: getDynamicConfigNumber,
  getBoolean: getDynamicConfigBoolean,
  getArray: getDynamicConfigArray,
  getObject: getDynamicConfigObject,
  addListener: addDynamicConfigListener,
  refresh: refreshDynamicConfig,
  clearCache: clearDynamicConfigCache,
  getStats: getDynamicConfigStats,
} = dynamicConfig;

export default dynamicConfig;

File: backend/core/dynamic_config.py
"""
Dinamik Konfigürasyon Sistemi
Runtime'da değiştirilebilir ayarlar
"""

import json
import os
from typing import Dict, Any, Optional, List
from datetime import datetime, timedelta
from dataclasses import dataclass, asdict
from core.config import settings
from core.logging import get_logger
from core.exceptions import ConfigurationError

logger = get_logger('dynamic_config')

@dataclass
class DynamicConfigItem:
    """Dinamik konfigürasyon öğesi"""
    key: str
    value: Any
    category: str
    description: str
    data_type: str
    is_sensitive: bool = False
    is_readonly: bool = False
    min_value: Optional[Any] = None
    max_value: Optional[Any] = None
    allowed_values: Optional[List[Any]] = None
    updated_at: datetime = None
    updated_by: Optional[str] = None

class DynamicConfigManager:
    """Dinamik konfigürasyon yöneticisi"""
    
    def __init__(self):
        self.config_file = "dynamic_config.json"
        self.config: Dict[str, DynamicConfigItem] = {}
        self.listeners: List[callable] = []
        self.load_config()
    
    def load_config(self) -> None:
        """Konfigürasyonu dosyadan yükle"""
        try:
            if os.path.exists(self.config_file):
                with open(self.config_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    
                for key, item_data in data.items():
                    # Convert datetime strings back to datetime objects
                    if item_data.get('updated_at'):
                        item_data['updated_at'] = datetime.fromisoformat(item_data['updated_at'])
                    
                    self.config[key] = DynamicConfigItem(**item_data)
                    
                logger.info(f"Dynamic config loaded: {len(self.config)} items")
            else:
                self.initialize_default_config()
                
        except Exception as e:
            logger.error(f"Failed to load dynamic config: {e}")
            self.initialize_default_config()
    
    def save_config(self) -> None:
        """Konfigürasyonu dosyaya kaydet"""
        try:
            data = {}
            for key, item in self.config.items():
                item_dict = asdict(item)
                # Convert datetime to string for JSON serialization
                if item_dict.get('updated_at'):
                    item_dict['updated_at'] = item_dict['updated_at'].isoformat()
                data[key] = item_dict
            
            with open(self.config_file, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
                
            logger.info("Dynamic config saved")
            
        except Exception as e:
            logger.error(f"Failed to save dynamic config: {e}")
            raise ConfigurationError(f"Failed to save dynamic config: {str(e)}")
    
    def initialize_default_config(self) -> None:
        """Varsayılan konfigürasyonu oluştur"""
        default_configs = [
            # OCR Settings
            DynamicConfigItem(
                key="ocr.parallel_workers",
                value=10,
                category="ocr",
                description="Paralel OCR işçi sayısı",
                data_type="integer",
                min_value=1,
                max_value=50
            ),
            DynamicConfigItem(
                key="ocr.timeout",
                value=30,
                category="ocr",
                description="OCR zaman aşımı (saniye)",
                data_type="integer",
                min_value=5,
                max_value=300
            ),
            DynamicConfigItem(
                key="ocr.confidence_threshold",
                value=0.7,
                category="ocr",
                description="OCR güven eşiği",
                data_type="float",
                min_value=0.1,
                max_value=1.0
            ),
            
            # Upload Settings
            DynamicConfigItem(
                key="upload.max_file_size_mb",
                value=10,
                category="upload",
                description="Maksimum dosya boyutu (MB)",
                data_type="integer",
                min_value=1,
                max_value=100
            ),
            DynamicConfigItem(
                key="upload.max_file_count",
                value=500,
                category="upload",
                description="Maksimum dosya sayısı",
                data_type="integer",
                min_value=1,
                max_value=1000
            ),
            DynamicConfigItem(
                key="upload.allowed_extensions",
                value=["jpg", "jpeg", "png", "webp"],
                category="upload",
                description="İzin verilen dosya uzantıları",
                data_type="list"
            ),
            
            # Cache Settings
            DynamicConfigItem(
                key="cache.ttl",
                value=300,
                category="cache",
                description="Cache TTL (saniye)",
                data_type="integer",
                min_value=60,
                max_value=3600
            ),
            DynamicConfigItem(
                key="cache.max_size_mb",
                value=100,
                category="cache",
                description="Maksimum cache boyutu (MB)",
                data_type="integer",
                min_value=10,
                max_value=1000
            ),
            
            # Performance Settings
            DynamicConfigItem(
                key="performance.slow_query_threshold",
                value=2.0,
                category="performance",
                description="Yavaş sorgu eşiği (saniye)",
                data_type="float",
                min_value=0.1,
                max_value=10.0
            ),
            DynamicConfigItem(
                key="performance.max_concurrent_requests",
                value=100,
                category="performance",
                description="Maksimum eşzamanlı istek",
                data_type="integer",
                min_value=10,
                max_value=1000
            ),
            
            # Security Settings
            DynamicConfigItem(
                key="security.max_login_attempts",
                value=5,
                category="security",
                description="Maksimum giriş denemesi",
                data_type="integer",
                min_value=3,
                max_value=10
            ),
            DynamicConfigItem(
                key="security.lockout_duration_minutes",
                value=15,
                category="security",
                description="Kilitlenme süresi (dakika)",
                data_type="integer",
                min_value=5,
                max_value=60
            ),
            
            # Feature Flags
            DynamicConfigItem(
                key="features.social_media",
                value=True,
                category="features",
                description="Sosyal medya özelliği",
                data_type="boolean"
            ),
            DynamicConfigItem(
                key="features.templates",
                value=True,
                category="features",
                description="Şablon özelliği",
                data_type="boolean"
            ),
            DynamicConfigItem(
                key="features.ai_generation",
                value=True,
                category="features",
                description="AI üretim özelliği",
                data_type="boolean"
            ),
            DynamicConfigItem(
                key="features.multi_language",
                value=False,
                category="features",
                description="Çoklu dil özelliği",
                data_type="boolean"
            ),
            
            # UI Settings
            DynamicConfigItem(
                key="ui.theme",
                value="light",
                category="ui",
                description="UI teması",
                data_type="string",
                allowed_values=["light", "dark", "system"]
            ),
            DynamicConfigItem(
                key="ui.language",
                value="tr",
                category="ui",
                description="UI dili",
                data_type="string",
                allowed_values=["tr", "en"]
            ),
            DynamicConfigItem(
                key="ui.date_format",
                value="DD/MM/YYYY",
                category="ui",
                description="Tarih formatı",
                data_type="string"
            ),
        ]
        
        for config_item in default_configs:
            config_item.updated_at = datetime.now()
            self.config[config_item.key] = config_item
        
        self.save_config()
        logger.info("Default dynamic config initialized")
    
    def get(self, key: str, default: Any = None) -> Any:
        """Konfigürasyon değerini al"""
        if key in self.config:
            return self.config[key].value
        return default
    
    def set(self, key: str, value: Any, updated_by: Optional[str] = None) -> bool:
        """Konfigürasyon değerini ayarla"""
        if key not in self.config:
            logger.warning(f"Unknown config key: {key}")
            return False
        
        config_item = self.config[key]
        
        # Check if readonly
        if config_item.is_readonly:
            logger.warning(f"Config key is readonly: {key}")
            return False
        
        # Validate value
        if not self.validate_value(config_item, value):
            return False
        
        # Update value
        old_value = config_item.value
        config_item.value = value
        config_item.updated_at = datetime.now()
        config_item.updated_by = updated_by
        
        # Save config
        self.save_config()
        
        # Notify listeners
        self.notify_listeners(key, old_value, value)
        
        logger.info(f"Config updated: {key} = {value}")
        return True
    
    def validate_value(self, config_item: DynamicConfigItem, value: Any) -> bool:
        """Değeri doğrula"""
        try:
            # Type validation
            if config_item.data_type == "integer":
                if not isinstance(value, int):
                    value = int(value)
            elif config_item.data_type == "float":
                if not isinstance(value, float):
                    value = float(value)
            elif config_item.data_type == "boolean":
                if not isinstance(value, bool):
                    value = bool(value)
            elif config_item.data_type == "string":
                if not isinstance(value, str):
                    value = str(value)
            elif config_item.data_type == "list":
                if not isinstance(value, list):
                    value = list(value)
            
            # Range validation
            if config_item.min_value is not None and value < config_item.min_value:
                logger.error(f"Value below minimum: {key} = {value} < {config_item.min_value}")
                return False
            
            if config_item.max_value is not None and value > config_item.max_value:
                logger.error(f"Value above maximum: {key} = {value} > {config_item.max_value}")
                return False
            
            # Allowed values validation
            if config_item.allowed_values is not None and value not in config_item.allowed_values:
                logger.error(f"Value not in allowed values: {key} = {value}")
                return False
            
            return True
            
        except (ValueError, TypeError) as e:
            logger.error(f"Value validation failed: {key} = {value}, error: {e}")
            return False
    
    def get_category(self, category: str) -> Dict[str, Any]:
        """Kategoriye göre konfigürasyonları al"""
        result = {}
        for key, config_item in self.config.items():
            if config_item.category == category:
                result[key] = {
                    'value': config_item.value,
                    'description': config_item.description,
                    'data_type': config_item.data_type,
                    'is_sensitive': config_item.is_sensitive,
                    'is_readonly': config_item.is_readonly,
                    'updated_at': config_item.updated_at,
                    'updated_by': config_item.updated_by
                }
        return result
    
    def get_all(self) -> Dict[str, Any]:
        """Tüm konfigürasyonları al"""
        result = {}
        for key, config_item in self.config.items():
            result[key] = {
                'value': config_item.value,
                'category': config_item.category,
                'description': config_item.description,
                'data_type': config_item.data_type,
                'is_sensitive': config_item.is_sensitive,
                'is_readonly': config_item.is_readonly,
                'updated_at': config_item.updated_at,
                'updated_by': config_item.updated_by
            }
        return result
    
    def add_listener(self, listener: callable) -> None:
        """Konfigürasyon değişiklik dinleyicisi ekle"""
        self.listeners.append(listener)
    
    def remove_listener(self, listener: callable) -> None:
        """Konfigürasyon değişiklik dinleyicisi kaldır"""
        if listener in self.listeners:
            self.listeners.remove(listener)
    
    def notify_listeners(self, key: str, old_value: Any, new_value: Any) -> None:
        """Dinleyicilere bildirim gönder"""
        for listener in self.listeners:
            try:
                listener(key, old_value, new_value)
            except Exception as e:
                logger.error(f"Config listener error: {e}")
    
    def reset_to_default(self, key: str) -> bool:
        """Konfigürasyonu varsayılan değere sıfırla"""
        if key not in self.config:
            return False
        
        # Reload default config to get original value
        original_config = DynamicConfigManager()
        original_config.initialize_default_config()
        
        if key in original_config.config:
            default_value = original_config.config[key].value
            return self.set(key, default_value, "system")
        
        return False
    
    def export_config(self) -> str:
        """Konfigürasyonu dışa aktar"""
        return json.dumps(self.get_all(), indent=2, ensure_ascii=False, default=str)
    
    def import_config(self, config_data: str) -> bool:
        """Konfigürasyonu içe aktar"""
        try:
            data = json.loads(config_data)
            for key, value in data.items():
                if key in self.config:
                    self.set(key, value, "import")
            return True
        except Exception as e:
            logger.error(f"Config import failed: {e}")
            return False

# Global dynamic config manager
dynamic_config = DynamicConfigManager()

def get_dynamic_config(key: str, default: Any = None) -> Any:
    """Dinamik konfigürasyon değerini al"""
    return dynamic_config.get(key, default)

def set_dynamic_config(key: str, value: Any, updated_by: Optional[str] = None) -> bool:
    """Dinamik konfigürasyon değerini ayarla"""
    return dynamic_config.set(key, value, updated_by)

def get_dynamic_config_category(category: str) -> Dict[str, Any]:
    """Kategoriye göre dinamik konfigürasyonları al"""
    return dynamic_config.get_category(category)

def get_all_dynamic_config() -> Dict[str, Any]:
    """Tüm dinamik konfigürasyonları al"""
    return dynamic_config.get_all()
